nilearn/decomposition/base.py

```diff
@@ -0,0 +1,461 @@
+
+    Uses a PCA (randomized for small reduction ratio) or a range finding matrix
+    on time series to reduce data size in time. For multiple image,
+    the concatenation of data is returned, either as an ndarray or a memorymap
+    (useful for big datasets that do not fit in memory).
+
+    Parameters
+    ----------
+    masker: NiftiMasker or MultiNiftiMasker
+        Instance used to mask provided data.
+
+    imgs: list of Niimg-like objects
+        See http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.
+        List of subject data.
+
+    confounds: CSV file path or 2D matrix
+        This parameter is passed to signal.clean. Please see the
+        corresponding documentation for details.
+
+    reduction_ratio: 'auto' or float in [0., 1.], optional
+        - Between 0. or 1. : controls compression of data, 1. means no
+        compression
+        - if set to 'auto', estimator will set the number of components per
+        compressed session to be n_components.
+
+    n_components: integer, optional
+        Number of components to be extracted by the PCA.
+
+    random_state: int or RandomState
+        Pseudo number generator state used for random sampling.
+
+    memory_level: integer, optional
+        Integer indicating the level of memorization. The higher, the more
+        function calls are cached.
+
+    memory: joblib.Memory
+        Used to cache the function calls.
+
+    Returns
+    ------
+    data: ndarray or memorymap
+        Concatenation of reduced data.
+    """
+
+    if not hasattr(imgs, '__iter__'):
+        imgs = [imgs]
+
+    if reduction_ratio == 'auto':
+        if n_components is None:
+            # Reduction ratio is 1 if
+            # neither n_components nor ratio is provided
+            reduction_ratio = 1
+    else:
+        if reduction_ratio is None:
+            reduction_ratio = 1
+        else:
+            reduction_ratio = float(reduction_ratio)
+        if not 0 <= reduction_ratio <= 1:
+            raise ValueError('Reduction ratio should be between 0., 1.,'
+                             'got %.2f' % reduction_ratio)
+
+    if confounds is None:
+        confounds = [None] * len(imgs)
+
+    # Precomputing number of samples for preallocation
+    subject_n_samples = np.zeros(len(imgs), dtype='int')
+    for i, img in enumerate(imgs):
+        this_n_samples = check_niimg_4d(img).shape[3]
+        if reduction_ratio == 'auto':
+            subject_n_samples[i] = min(n_components,
+                                       this_n_samples)
+        else:
+            subject_n_samples[i] = int(ceil(this_n_samples *
+                                            reduction_ratio))
+    n_voxels = np.sum(_safe_get_data(masker.mask_img_))
+    n_samples = np.sum(subject_n_samples)
+
+    # XXX Should we provided memory mapping for n_jobs > 1 to allow concurrent
+    # write ?
+    data = np.empty((n_samples, n_voxels), order='F',
```

Comment:
I think that we should move this affection below, as it is not used before the for loop (useful for readability and memory usage).